File: libs/core/src/private/symbols/parameter_symbol.cpp
/**
 * @file parameter_symbol.cpp
 * @author Francesco Corso
 * @date 8/8/2026
 * @brief
 */
module prism.core:symbols.parameter_symbol.impl;

import :symbols.parameter_symbol;
import :symbols.type_symbol;

namespace prism
{

    void ParameterSymbol::write_display_string(TextWriter &writer) const
    {
        writer.write(name().as_string_view());
        writer.write(": ");
        type().write_display_string(writer);
    }
} // namespace prism

File: src/Prism.Core/Symbols/ParameterSymbol.cs
using System.Diagnostics.CodeAnalysis;
using Prism.Core.Semantic;
using Prism.Core.Syntax;

namespace Prism.Core.Symbols;

public closed record ParameterDefault(SyntaxNode? Syntax);

public sealed record ConstantParameterDefault(ConstantValue Value, SyntaxNode? Syntax = null)
    : ParameterDefault(Syntax);

public closed class ParameterSymbol : ValueSymbol
{
    private protected ParameterSymbol(string name, Symbol? containingSymbol)
        : base(name, containingSymbol) { }

    public abstract FunctionSymbol? ContainingFunction { get; }

    [MemberNotNullWhen(true, nameof(DefaultValue))]
    public bool HasDefaultValue => DefaultValue is not null;

    public abstract ParameterDefault? DefaultValue { get; }

    public sealed override void WriteDisplayString(TextWriter writer)
    {
        writer.Write(Name);
        writer.Write(": ");
        Type.WriteDisplayString(writer);
    }
}
